src/traffic_server/InkAPI.cc

```diff
@@ -1333,6 +1349,103 @@ APIHooks::clear()
   }
 }
 
+HttpHookState::HttpHookState() : _id(TS_HTTP_LAST_HOOK) {}
+
+void
+HttpHookState::init(TSHttpHookID id, HttpAPIHooks const *global, HttpAPIHooks const *ua, HttpAPIHooks const *sm)
+{
+  _id = id;
+
+  if (global) {
+    _global.init(global, id);
+  } else {
+    _global.clear();
+  }
+
+  if (ua) {
+    _ssn.init(ua, id);
+  } else {
+    _ssn.clear();
+  }
+
+  if (sm) {
+    _txn.init(sm, id);
+  } else {
+    _txn.clear();
+  }
+}
+
+APIHook const *
+HttpHookState::getNext()
+{
+  APIHook const *zret = nullptr;
+  do {
+    APIHook const *hg   = _global.candidate();
+    APIHook const *hssn = _ssn.candidate();
+    APIHook const *htxn = _txn.candidate();
+    zret                = nullptr;
+
+    Debug("plugin", "computing next callback for hook %d", _id);
+
+    if (hg) {
+      zret = hg;
+      ++_global;
+    } else if (hssn) {
+      zret = hssn;
+      ++_ssn;
+    } else if (htxn) {
+      zret = htxn;
+      ++_txn;
+    }
+  } while (zret != nullptr && !this->is_enabled());
+
+  return zret;
+}
+
+bool
+HttpHookState::is_enabled()
+{
+  return true;
+}
+
+void
+HttpHookState::Scope::init(HttpAPIHooks const *feature_hooks, TSHttpHookID id)
+{
+  APIHooks const *hooks = (*feature_hooks)[id];
+
+  _c = _p = nullptr;
+
+  if (hooks) {
+    _c = hooks->head();
+  }
+}
+
+APIHook const *
+HttpHookState::Scope::candidate()
+{
+  APIHook const *x = nullptr;
+  if (nullptr != _c) {
```

Comment:
This needs some comments. The logic is not obvious.

Reply:
Since in this commit, we won't have any hook insertion based on priority, maybe we should just skip this while loop?